src/memoize-decorator.ts

```diff
@@ -1,62 +1,92 @@
-export function Memoize(autoHashOrHashFn?: boolean | ((...args: any[]) => any)) {
-	return (target: Object, propertyKey: string, descriptor: TypedPropertyDescriptor<any>) => {
+interface MemoizeArgs {
+	expiring?: number;
+	hashFunction?: boolean | ((...args: any[]) => any);
+	tags?: string[];
+}
+
+export function Memoize(args?: MemoizeArgs | MemoizeArgs['hashFunction']) {
+	let hashFunction: MemoizeArgs['hashFunction'];
+	let duration: MemoizeArgs['expiring'];
+	let tags: MemoizeArgs['tags'];
 
+	if (typeof args === 'object') {
+		hashFunction = args.hashFunction;
+		duration = args.expiring;
+		tags = args.tags;
+	} else {
+		hashFunction = args;
+	}
+
+	return (target: Object, propertyKey: string, descriptor: TypedPropertyDescriptor<any>) => {
 		if (descriptor.value != null) {
-			descriptor.value = getNewFunction(descriptor.value, autoHashOrHashFn);
+			descriptor.value = getNewFunction(descriptor.value, hashFunction, duration, tags);
 		} else if (descriptor.get != null) {
-			descriptor.get = getNewFunction(descriptor.get, autoHashOrHashFn);
+			descriptor.get = getNewFunction(descriptor.get, hashFunction, duration, tags);
 		} else {
 			throw 'Only put a Memoize() decorator on a method or get accessor.';
 		}
-
 	};
 }
 
-export function MemoizeExpiring(duration: number, autoHashOrHashFn?: boolean | ((...args: any[]) => any)) {
-	return (target: Object, propertyKey: string, descriptor: TypedPropertyDescriptor<any>) => {
+export function MemoizeExpiring(expiring: number, hashFunction?: MemoizeArgs['hashFunction']) {
+	return Memoize({
+		expiring,
+		hashFunction
+	});
+}
 
-		if (descriptor.value != null) {
-			descriptor.value = getNewFunction(descriptor.value, autoHashOrHashFn, duration);
-		} else if (descriptor.get != null) {
-			descriptor.get = getNewFunction(descriptor.get, autoHashOrHashFn, duration);
-		} else {
-			throw 'Only put a Memoize() decorator on a method or get accessor.';
-		}
+const clearCacheTagsMap: Map<string, Map<any, any>[]> = new Map();
 
-	};
+export function clear (tags: string[]): number {
+	const cleared: Set<Map<any, any>> = new Set();
+	for (const tag of tags) {
+		const maps = clearCacheTagsMap.get(tag);
+		for (const mp of maps) {
+			if (!cleared.has(mp)) {
+				mp.clear();
+				cleared.add(mp);
+			}
+		}
```

Comment:
Can be replaced with:
```
maps?.forEach((mp) => {
	if (!cleared.has(mp)) {
		mp.clear();
		cleared.add(mp);
	}
});
```

Reply:
Hi, I have done the fix, and added the case to be verified to the test. I kept using "for ... of" because it should perform better than the forEach callback.